### DISCUSSION TITLE
Loading multiple time-series for feature extraction?

### DISCUSSION THREAD

**Discussion by ddofer** [alan-turing-institute/sktime#2823] [Q&A]
I'm trying to extract features for multiple different time-series. I'm having difficulty finding how to transform the data to a compatible format. 
Currently, the time-series are univariate, although I would like to eventually apply this to multiple, multivariate TS. 
The website documentation link to [AA notebooks is broken](https://notebooks.githubusercontent.com/view/ipynb?browser=chrome&color_mode=auto&commit=727789da8d86821576b467c119ff07fb11059f94&device=unknown&enc_url=68747470733a2f2f7261772e67697468756275736572636f6e74656e742e636f6d2f616c616e2d747572696e672d696e737469747574652f736b74696d652f373237373839646138643836383231353736623436376331313966663037666231313035396639342f6578616d706c65732f41415f6461746174797065735f616e645f64617461736574732e6970796e62&logged_in=false&nwo=alan-turing-institute%2Fsktime&path=examples%2FAA_datatypes_and_datasets.ipynb&platform=android&repository_id=156401841&repository_type=Repository&version=101), and said notebooks doesn't provide the appropiate format, for, e.g. catch22 or TSFresh feature extraction. 
Thanks!

Data looks like this ("name" is the column defining each individual time series, if needed I can pivot? - It seems to be in the pd_multiindex format as defined? )
![image](https://user-images.githubusercontent.com/5120387/174045165-99fdbc98-970e-4faf-a976-5fe8a76f266e.png)
`c22 = Catch22()
c22.fit_transform(data)`
Results in error:

> TypeError: X must be in an sktime compatible format, of scitype Series, Panel or Hierarchical, for instance a pandas.DataFrame with sktime compatible time indices, or with MultiIndex and lowest level a sktime compatible time index. allowed compatible mtype format specifications are: ['pd.Series', 'pd.DataFrame', 'np.ndarray', 'nested_univ', 'numpy3D', 'pd-multiindex', 'df-list', 'pd_multiindex_hier'] See the data format tutorial examples/AA_datatypes_and_datasets.ipynb

**Comment by fkiraly** [comment]
Can you be more specific how/why you think the link to the datatypes notebook is broken, and where? I follow your link and it shows the correct notebook.

> It seems to be in the pd_multiindex format as defined?

If you run `check_raise(data, "pd-multiindex")`, it should tell you that the last multiindex level should be time-like, but in your case it is the first. What shouhld do the trick is:

`data = data.swaplevel()` with the end state

or doing `data.set_index(["name", "year"])` instead, earlier on

**Reply by fkiraly** [reply]
> said notebooks doesn't provide the appropiate format, for, e.g. catch22 or TSFresh feature extraction.

The format to use in those feature extractors should be as stated in the notebook, can you explain why you think it does not? So we can fix this.

**Reply by ddofer** [reply]
1. I tried "Reversing" the set_index order, still results in the same error. 
![image](https://user-images.githubusercontent.com/5120387/174055850-90541845-68b5-4650-8dfd-09ca3141d3de.png)

Re error - the link in the "Getting started" documentation is missing - https://www.sktime.org/en/stable/examples/loading_data.html

**Reply by ddofer** [reply]
i.e that doesn't fix it

**Reply by fkiraly** [reply]
> the link in the "Getting started" documentation is missing

Ah - known issue, https://github.com/alan-turing-institute/sktime/issues/2127
someone should work on documentation... we´ll have a doc sprint in Aug, contributions appreciated.

**Reply by fkiraly** [reply]
>that doesn't fix it

Looks correct to me though - what do you get if you run `check_raise`?

**Reply by fkiraly** [reply]
so, based on the 4 hours of silence, I assume your computer exploded as soon as you ran `check_raise`.

If that's the case, kindly open a bug report with `sktime` and dependency versions, operating system, computer manufacturer, and explosion radius.

**Comment by fkiraly** [comment]
This indicates our error messages might be suboptimal.
Made a "quick fix" and opened a refactor/improvement issue.

https://github.com/alan-turing-institute/sktime/pull/2824
https://github.com/alan-turing-institute/sktime/issues/2825

PS @ddofer, you may like to look at the new error message in #2825.

This points to `sktime.datatypes.check_raise` which you can use to diagnose incompatible inputs.
Is that more helpful?

**Reply by ddofer** [reply]
More essential is a clearer example of what the format is expected to be, e.g. `multiindex = pd.dataframe where index is the identifier of each series and the date col, and remaining columns are the values per series`

**Reply by fkiraly** [reply]
that´s in the AA tutorial, I thought

**Reply by fkiraly** [reply]
![image](https://user-images.githubusercontent.com/7985502/174057831-faaba152-7118-4f56-b14b-6286c57c2946.png)

**Reply by fkiraly** [reply]
let me know why that was not good here, e.g., what should be improved

**Reply by ddofer** [reply]
Do I need to convert it, from a multiindex dataframe?

**Reply by ddofer** [reply]
i.e., My data is a multiindex dataframe (of multiple series), what do I need to do further, in order to extract features per time series with the sktime framework extractors? e.g. catch22, tsfresh? 
Thanks again!

**Reply by fkiraly** [reply]
just apply the transformer to the data (`fit_transform`). It should create a `pd.DataFrame` where each row corresponds to one time series.

**Comment by fkiraly** [comment]
Oh, also the error message is wrong. "lowest level" is the one on the left, is that correct? It should be the one on the right, the "last level", that is timelike.

**Reply by fkiraly** [reply]
"lowest" is common terminology in hierarchical time series, if you think about tree/nodes, but I see how this can be very ambiguous, since the "lowest" level in that sense has the highest integer index in `pandas`. Changed it to "last" to clarify this, hope the message is better now:

```
            for instance a pandas.DataFrame with sktime compatible time indices, 
            or with MultiIndex and last (-1) level a sktime compatible time index.
```

better?
